pkg/controller/statefulset/stateful_pod_control.go

```diff
@@ -135,15 +191,97 @@ func (spc *realStatefulPodControl) UpdateStatefulPod(set *apps.StatefulSet, pod
 	return err
 }
 
-func (spc *realStatefulPodControl) DeleteStatefulPod(set *apps.StatefulSet, pod *v1.Pod) error {
-	err := spc.client.CoreV1().Pods(set.Namespace).Delete(context.TODO(), pod.Name, metav1.DeleteOptions{})
+func (spc *StatefulPodControl) DeleteStatefulPod(set *apps.StatefulSet, pod *v1.Pod) error {
+	err := spc.objectMgr.DeletePod(pod)
 	spc.recordPodEvent("delete", set, pod, err)
 	return err
 }
 
+// ClaimsMatchDeletionPolicy returns false if the PVCs for pod are not consistent with set's PVC deletion policy.
+// An error is returned if something is not consistent. This is expected if the pod is being otherwise updated,
+// but a problem otherwise (see usage of this method in UpdateStatefulPod).
+func (spc *StatefulPodControl) ClaimsMatchDeletionPolicy(set *apps.StatefulSet, pod *v1.Pod) (bool, error) {
+	ordinal := getOrdinal(pod)
+	templates := set.Spec.VolumeClaimTemplates
+	policy := set.Spec.PersistentVolumeClaimDeletePolicy
+	claimShouldBeRetained := policy == apps.Retain || policy == apps.DeleteOnStatefulSetDeletionOnly
+	for i := range templates {
+		claimName := getPersistentVolumeClaimName(set, &templates[i], ordinal)
+		claim, err := spc.objectMgr.GetClaim(templates[i].Namespace, claimName)
+		switch {
+		case apierrors.IsNotFound(err):
```

Comment:
This also happens when everything is getting created IIUC. In UpdateStatefulPod, above, we ignore an error from this method when appropriate.

I think it's nicer to do it that way, where the caller knows when it's ok to ignore missing things, than to plumb that logic down here.

I may eat my words on this when we do more complicated test cases, but for the time being at least to me it makes it easier to reason about.

Reply:
If objectMgr is fed by an informer, this in an inherently racy check (create object A, check cache until B shows up).  Generally we don't warn in those cases, because that's creating log spam (the race is roughly guaranteed to always happen at least once on the first pass, and depending on watch latency, possibly longer).

If a PVC doesn't exist, retention is irrelevant.  So I'd probably say apierrors.IsNotFound should not cause an error to be returned, and should log at a high level (i.e. not be visible except during debugging such as 4 or 5).  It doesn't matter whether it's a race on startup, a user driven deletion, or a bug elsewhere in the controller - the controller has to accept that cross object comparisons driven by inherently laggy caches can't be done transactionally, and the way we do it is break our checks into simple invariants (if no PVC exists but should -> try to create, if PVC shouldn't exist but does -> try to destroy) and minimizing the amount of noise we create in logs and events.

Reply:
I think that makes sense. Done.